src/wallet/mod.rs

```diff
@@ -366,67 +332,236 @@ where
+
+            self.spk_tracker.add_spk((keychain, index), spk);
+        }
+
+        true
     }
 
     /// Return whether or not a `script` is part of this wallet (either internal or external)
-    pub fn is_mine(&self, script: &Script) -> Result<bool, Error> {
-        self.database.borrow().is_mine(script)
+    pub fn is_mine(&self, script: &Script) -> bool {
+        self.spk_tracker.index_of_spk(script).is_some()
     }
 
     /// Return the list of unspent outputs of this wallet
     ///
     /// Note that this method only operates on the internal database, which first needs to be
     /// [`Wallet::sync`] manually.
-    pub fn list_unspent(&self) -> Result<Vec<LocalUtxo>, Error> {
-        Ok(self
-            .database
-            .borrow()
-            .iter_utxos()?
-            .into_iter()
-            .filter(|l| !l.is_spent)
-            .collect())
+    pub fn list_unspent(&self) -> Vec<LocalUtxo> {
+        self.spk_tracker
+            .iter_unspent(&self.sparse_chain, &self.tx_graph)
+            .map(|((keychain, _), txo)| LocalUtxo {
+                outpoint: txo.outpoint,
+                txout: txo.txout,
+                keychain,
+                is_spent: false,
+            })
+            .collect()
+    }
+
+    /// Iterate over all checkpoints.
+    pub fn iter_checkpoints(&self) -> &BTreeMap<u32, BlockHash> {
+        self.sparse_chain.checkpoints()
+    }
+
+    /// Range checkpoints
+    pub fn range_checkpoints<'a>(
+        &'a self,
+        range: impl RangeBounds<u32> + 'a,
+    ) -> impl DoubleEndedIterator<Item = BlockId> + 'a {
+        self.sparse_chain.range_checkpoints(range)
+    }
+
+    /// Returns the latest checkpoint.
+    pub fn latest_checkpoint(&self) -> Option<BlockId> {
+        self.sparse_chain.latest_checkpoint()
+    }
+
+    /// Inserts a transaction into the graph if the transaction is deemed relevant.
+    /// Also enforces the stop_gap.
+    pub fn graph_insert_if_relevant(&mut self, tx: &Transaction, stop_gap: Option<u32>) -> bool {
+        let mut is_relevant = false;
+
+        for keychain in [KeychainKind::External, KeychainKind::Internal] {
+            let spk_range = (keychain, u32::MIN)..(keychain, u32::MAX);
+
+            match self
+                .spk_tracker
+                .highest_spk_index(&self.tx_graph, tx, spk_range)
+            {
+                Some((_, derivation)) => {
+                    if let Some(stop_gap) = stop_gap {
+                        let max_addresses = derivation.saturating_add(stop_gap);
+                        self.ensure_derived_up_to(keychain, max_addresses);
+                    }
+                }
+                None => continue,
+            }
+
+            // skip if already in chain
+            // we check chain instead of graph, as graph may be updated without commiting to chain
+            if self.sparse_chain.transaction_height(tx.txid()).is_some() {
+                continue;
+            }
```

Comment:
This checks seems relatively cheap, maybe it should be done first, before the match?